quinn-proto/src/bytes_source.rs

```diff
@@ -0,0 +1,231 @@
+    ///
+    /// This method will consume parts of the source.
+    // Calling it will yield `Bytes` elements up to the configured `limit`.
+    fn pop_chunk(&mut self) -> Option<Bytes>;
+}
+
+/// Indicates how many bytes and chunks had been transferred in a write operation
+#[derive(Debug, Default, PartialEq, Eq, Clone, Copy)]
+pub struct Written {
+    /// The amount of bytes which had been written
+    pub bytes: usize,
+    /// The amount of full chunks which had been written
+    ///
+    /// If a chunk was only partially written, it will not be counted by this field.
+    pub chunks: usize,
+}
+
+/// A [`BytesSource`] implementation for `&'a mut [Bytes]`
+///
+/// The type allows to dequeue [`Bytes`] chunks from an array of chunks, up to
+/// a configured limit.
+pub struct BytesArray<'a> {
+    /// The wrapped slice of `Bytes`
+    chunks: &'a mut [Bytes],
+    /// The amount of bytes which are still available for consumption
+    len: usize,
+    /// The amount of bytes and chunks consumed from this source
+    consumed: Written,
+}
+
+impl<'a> BytesArray<'a> {
+    pub fn from_chunks(chunks: &'a mut [Bytes]) -> Self {
+        let len = chunks.iter().map(|chunk| chunk.len()).sum();
+        Self {
+            chunks,
+            len,
+            consumed: Written::default(),
+        }
+    }
+}
+
+impl<'a> BytesSource for BytesArray<'a> {
+    fn len(&self) -> usize {
+        self.len
+    }
+
+    fn limit(&mut self, new_len: usize) {
+        self.len = self.len.min(new_len);
+    }
+
+    fn consumed(&self) -> Written {
+        self.consumed
+    }
+
+    fn pop_chunk(&mut self) -> Option<Bytes> {
+        // The loop exists to skip empty chunks while still marking them as
+        // consumed
+        while self.consumed.chunks < self.chunks.len() {
+            let chunk = &mut self.chunks[self.consumed.chunks];
+
+            if chunk.len() <= self.len {
+                let chunk = std::mem::take(chunk);
+                self.consumed.chunks += 1;
+                if chunk.is_empty() {
+                    continue;
+                }
+                self.len -= chunk.len();
+                self.consumed.bytes += chunk.len();
+                return Some(chunk);
+            } else if self.len > 0 {
+                let chunk = chunk.split_to(self.len);
+                self.len = 0;
+                self.consumed.bytes += chunk.len();
+                return Some(chunk);
+            } else {
+                return None;
+            }
+        }
+
+        None
+    }
+}
+
+/// A [`BytesSource`] implementation for `&[u8]`
+///
+/// The type allows to dequeue a single [`Bytes`] chunk, which will be lazily
+/// created from a reference. This allows to defer the allocation until it is
+/// known how much data needs to be copied.
+pub struct ByteSlice<'a> {
+    /// The wrapped byte slice
+    data: &'a [u8],
+    /// The amount of bytes which are still available for consumption
+    len: usize,
```

Comment:
There are 2 reasons why not:
1. The borrow checker hates me (I tried it!)
2. We would need a separate `limit` field in order to return later on whether the full chunk was written, or would need to additionally store the original data length without slicing. And while `Written::chunks` iis not terribly useful for the normal slice based write API since it is returned to nobody - I think it is still nice if everything upholds the API contract. Otherwise it might confuse people later on.